### DISCUSSION TITLE
NPM npmrc resolution appears to be order-sensitive

### DISCUSSION THREAD

**Discussion by aisamu** [renovatebot/renovate#17922] [Request Help]
We had some trouble with private registries on a monorepo that had multiple `package.json` files sharing a root `.npmrc`.
Renovate was using the correct registry on half of the packages, but on the other half it was incorrectly using the default npm registry.

If we say that the order on which the files are presented on the Dashboard `Detected Dependencies` section correlates with the order of "visiting" those files, we could say that the root `package.json` was the breaking point: all of the `package.json` resolutions before it had issues; all the ones after it worked.

After some investigation, it appears that Renovate's handling of `.npmrc` is indeed order-sensitive.
The following repository reproduces the scenario with a detailed explanation: https://github.com/Stedi/renovate-root-npmrc

This was a bit surprising, as we'd expect `.npmrc` files either to be truly global (applying to _all_ the nested folders regardless of the order, like `npm`) or perhaps to be isolated (applying only to sibling `package.json` files).

**Comment by rarkins** [comment]
As far as I'm aware, `npm` does not "hunt" up the file path to find `.npmrc` files. If you run `npm x` in directory `/a/b/c/d` then it will look for `/a/b/c/d/.npmrc` and not in `/a/b/c/.npmrc`, etc.

Demonstrating using your example:

```
❯ pwd
/tmp/renovate-root-npmrc/a-abritrary/arbitrary-a
❯ ls -al
total 8
drwxr-xr-x  3 rhys  wheel   96 Sep 25 06:03 .
drwxr-xr-x  3 rhys  wheel   96 Sep 25 06:03 ..
-rw-r--r--  1 rhys  wheel  132 Sep 25 06:03 package.json
❯ cat package.json
{
  "name": "arbitrary-a",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@nonexistent/package-a": "1.0.0"
  }
}
❯ npm i
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@nonexistent%2fpackage-a - Not found
npm ERR! 404
npm ERR! 404  '@nonexistent/package-a@1.0.0' is not in this registry.
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/rhys/.npm/_logs/2022-09-25T04_04_25_775Z-debug-0.log
```

So I think the incorrect part is that Renovate should not be applying the root `.npmrc` registry setting to the `z-arbitrary/arbitrary-z/package.json` lookup. Do you agree?

**Reply by aisamu** [reply]
> npm does not "hunt" up the file path to find .npmrc files. 

That's correct!

The extra context is that those are [npm workspaces](https://docs.npmjs.com/cli/v8/using-npm/workspaces), and one of the documented/default flows is to [invoke commands from the root folder, either targeting a specific workspace or all workspaces](https://docs.npmjs.com/cli/v8/using-npm/workspaces#running-commands-in-the-context-of-workspaces).

On the same repo, if we remove the top-level broken dependencies to reduce noise...
```diff
diff --git a/package.json b/package.json
index 66b091b..9db9ab9 100644
--- a/package.json
+++ b/package.json
@@ -17,8 +17,5 @@
     "a-arbitrary/*",
     "z-arbitrary/*"
   ],
-  "dependencies": {
-    "@nonexistent/package-a": "1",
-    "@nonexistent/root-package": "2"
-  }
+  "dependencies": {}
 }
 ```
...and run `npm i` on the root: 
```
 ❯ pwd        
/xxx/renovate-root-npmrc

❯ npm i                            
npm ERR! code E401
npm ERR! 401 Unauthorized - GET https://npm.pkg.github.com/@nonexistent%2fpackage-a - authentication token not provided
```

We can see that `npm` tried to install the dependencies _of the workspaces_ using the root `.npmrc`!

<details>
<summary>Versions</summary>
<pre>
❯ node --version && npm --version
v18.9.0
8.19.1
</pre>
</details>

> So I think the incorrect part is that Renovate should not be applying the root .npmrc registry setting to the z-arbitrary/arbitrary-z/package.json lookup. Do you agree?

Given the above, I'm not 100% sure.
Since the root `.npmrc` seems to be enough for the regular/usual workspace flow, mirroring this behavior might be the path of least (user) friction.

I do agree that, without workspaces, we should _also_ mirror `npm` and not use "dominating" `.npmrc`s

**Reply by rarkins** [reply]
So is the problem that Renovate is not acting as if it's running commands from the workspaces root folder?
